File: solutions/782/main.py
from typing import * 

class Solution:
    def movesToChessboard(self, board: List[List[int]]) -> int:
        n = len(board)
        # corners of every rectangle should be 0000, 1111 or 0011
        for r in range(n): 
            for c in range(n):
                if board[0][0] ^ board[r][0] ^ board[r][c] ^ board[0][c]:
                    return -1
        # traverse board 
        rowSum = colSum = rowSwap = colSwap = 0  
        for i in range(n):
            rowSum += board[0][i]
            colSum += board[i][0]
            rowSwap += (board[i][0] == i%2)
            colSwap += (board[0][i] == i%2)
        # check number of 0, 1 
        if rowSum != n//2 and rowSum != (n+1)//2:
            return -1 
        if colSum != n//2 and colSum != (n+1)//2:
            return -1 
        # check min step based on n even or odd
        if n % 2: 
            if colSwap % 2:
                colSwap = n - colSwap
            if rowSwap % 2: 
                rowSwap = n - rowSwap
        else: 
            colSwap = min(n - colSwap, colSwap)
            rowSwap = min(n - rowSwap, rowSwap)
        # /2 since we fix 2 nums at each step 
        return (rowSwap + colSwap) // 2  
 
if __name__ == "__main__": 
    s = Solution()

    board = [[0,1,1,0],[0,1,1,0],[1,0,0,1],[1,0,0,1]]
    assert s.movesToChessboard(board) == 2 

File: solutions/782/main.rs
use std::ops::Index;

// ref https://stackoverflow.com/a/64499219/8519188
fn transpose<T>(v: Vec<Vec<T>>) -> Vec<Vec<T>> {
    assert!(!v.is_empty());
    let len = v[0].len();
    let mut iters: Vec<_> = v.into_iter().map(|n| n.into_iter()).collect();
    (0..len)
        .map(|_| {
            iters
                .iter_mut()
                .map(|n| n.next().unwrap())
                .collect::<Vec<T>>()
        })
        .collect()
}

impl Solution {
    pub fn moves_to_chessboard(board: Vec<Vec<i32>>) -> i32 {
        let n = board.len();
        
        for r in 0..n {
            for c in 0..n {
                if (board[0][0] ^ board[r][0] ^ board[r][c] ^ board[0][c]) == 1 {
                    return -1; 
                }
            }
        }

        let add_i = |i, x| {
            if x as usize == i % 2 {
                return 1; 
            } else {
                return 0; 
            }
        };

        let (mut rowSum, mut colSum, mut rowSwap, mut colSwap) = (0, 0, 0, 0); 
        let rowSum = Index::index(&board, 0).iter().fold(0, |acc, x| acc + x); 
        let mut rowSwap = board[0].clone().into_iter().enumerate().fold(0, |acc, (i, x)| acc + add_i(i, x)); 
        let board_T = transpose(board); 
        let colSum = Index::index(&board_T, 0).iter().fold(0, |acc, x| acc + x);
        let mut colSwap = board_T[0].clone().into_iter().enumerate().fold(0, |acc, (i, x)| acc + add_i(i, x)); 
        
        if rowSum as usize != n/2 && rowSum as usize != (n+1)/2 {
            return -1; 
        }
        if colSum as usize != n/2 && colSum as usize != (n+1)/2 {
            return -1; 
        }

        if n % 2 == 1{
            if rowSwap % 2 == 1 {
                rowSwap = n - rowSwap; 
            }
            if colSwap % 2 == 1 {
                colSwap = n - colSwap; 
            }
        }
        else {
            colSwap = colSwap.min(n - colSwap);
            rowSwap = rowSwap.min(n - rowSwap);
        }
        ((rowSwap as i32 + colSwap as i32) / 2) 
    }
}